packages/ui/VirtualList/VirtualListBase.js

```diff
@@ -23,728 +27,769 @@ const gridListItemSizeShape = PropTypes.shape({
+
+		setContainerSize = () => {
+			if (this.contentRef) {
+				this.contentRef.style.width = this.scrollBounds.scrollWidth + 'px';
+				this.contentRef.style.height = this.scrollBounds.scrollHeight + 'px';
+			}
 		}
-	}
 
-	syncThreshold (maxPos) {
-		const {threshold} = this;
+		updateMoreInfo (dataSize, primaryPosition) {
+			const
+				{dimensionToExtent, moreInfo} = this,
+				{itemSize, gridSize, clientSize} = this.primary;
 
-		if (threshold.max > maxPos) {
-			if (maxPos < threshold.base) {
-				threshold.max = threshold.base;
-				threshold.min = -Infinity;
+			if (dataSize <= 0) {
+				moreInfo.firstVisibleIndex = null;
+				moreInfo.lastVisibleIndex = null;
 			} else {
-				threshold.max = maxPos;
-				threshold.min = maxPos - threshold.base;
+				moreInfo.firstVisibleIndex = (Math.floor((primaryPosition - itemSize) / gridSize) + 1) * dimensionToExtent;
+				moreInfo.lastVisibleIndex = Math.min(dataSize - 1, Math.ceil((primaryPosition + clientSize) / gridSize) * dimensionToExtent - 1);
 			}
 		}
-	}
-
-	setScrollPosition (x, y, dirX, dirY) {
-		const
-			{dataSize} = this.props,
-			{firstIndex, numOfItems} = this.state,
-			{isPrimaryDirectionVertical, threshold, dimensionToExtent, maxFirstIndex, scrollBounds} = this,
-			{gridSize} = this.primary,
-			maxPos = isPrimaryDirectionVertical ? scrollBounds.maxTop : scrollBounds.maxLeft,
-			minOfMax = threshold.base,
-			maxOfMin = maxPos - threshold.base;
-		let delta, numOfGridLines, newFirstIndex = firstIndex, pos, dir = 0;
-
-		if (isPrimaryDirectionVertical) {
-			pos = y;
-			dir = dirY;
-		} else {
-			pos = x;
-			dir = dirX;
-		}
 
-		if (dir === 1 && pos > threshold.max) {
-			delta = pos - threshold.max;
-			numOfGridLines = Math.ceil(delta / gridSize); // how many lines should we add
-			threshold.max = Math.min(maxPos, threshold.max + numOfGridLines * gridSize);
-			threshold.min = Math.min(maxOfMin, threshold.max - gridSize);
-			newFirstIndex += numOfGridLines * dimensionToExtent;
-		} else if (dir === -1 && pos < threshold.min) {
-			delta = threshold.min - pos;
-			numOfGridLines = Math.ceil(delta / gridSize);
-			threshold.max = Math.max(minOfMax, threshold.min - (numOfGridLines * gridSize - gridSize));
-			threshold.min = (threshold.max > minOfMax) ? threshold.max - gridSize : -Infinity;
-			newFirstIndex -= numOfGridLines * dimensionToExtent;
+		syncThreshold (maxPos) {
+			const {threshold} = this;
+
+			if (threshold.max > maxPos) {
+				if (maxPos < threshold.base) {
+					threshold.max = threshold.base;
+					threshold.min = -Infinity;
+				} else {
+					threshold.max = maxPos;
+					threshold.min = maxPos - threshold.base;
+				}
+			}
 		}
 
-		if (threshold.min === -Infinity) {
-			newFirstIndex = 0;
-		} else {
-			newFirstIndex = Math.min(maxFirstIndex, newFirstIndex);
-			newFirstIndex = Math.max(0, newFirstIndex);
+		// Native only
+		scrollToPosition (x, y) {
+			this.containerRef.scrollTo(
```

Comment:
Let's add a guard code for this.containerRef?

Reply:
👌